### PR TITLE
Add csv mode and delay parameter to pcm-iio

### PR DESCRIPTION
**Pull request by bigstream-dmytro**
Currently pcm-iio is able only to send output to cout and the output becomes quite unreadable in case when you want to see how some metric is changed in time. `-csv` option was added to make handle this case.
Also now there is a hardcoded measurement interval. The proposed change allows to pass the interval in milliseconds to the utility.

### CODE REVIEW

**Comment by opcm** [comment]
thanks for the patch!

**Comment by opcm** [comment]
> Globally commenting on the millisecond change:
> As far as I know all pcm programs use a floating point number for delay, so if you pass it 0.001 it should use 1 ms as delay. Not sure if that worked though.
> The option to have CSV output is much appreciated, but I will let others comment on how it is done.

That is right (delay in seconds as a floating point number)

**Comment by bigstream-dmytro** [comment]
> That is right (delay in seconds as a floating point number)

Changed to floating point.

**Review by ilia-kurakin — COMMENTED** [review]
@bigstream-dmytro, thanks a lot for your patch. I'm ok with it, let's just make csv format more natural for automatic parsing.

File: pcm-iio.cpp [hunk 1]
```diff
@@ -333,11 +333,61 @@ vector<string> build_display(vector<struct iio_stacks_on_socket>& iios, vector<s
...
+                vector<uint64_t> h_data;
+                string v_name = h_array[0]->v_event_name;
+                current_row.clear();
+                current_row.push_back(socket_name);
+                current_row.push_back(stack_name);
+                current_row.push_back(v_name);
+                for (map<uint32_t,struct counter*>::const_iterator hunit = h_array.cbegin(); hunit != h_array.cend(); ++hunit) {
+                    uint32_t hh_id = hunit->first;
+                    uint64_t raw_data = hunit->second->data[0][socket->socket_id][stack_id][std::pair<h_id,v_id>(hh_id,vv_id)];
+                    current_row.push_back(unit_format(raw_data));
```
[the remark this answers is not in the document]
Comment (bigstream-dmytro) [reply pcm-iio.cpp:377]: Sometimes csv is more convenient even as human-readable output. I suggest to introduce an extra parameter, e.g. `-human-readable`. In this mode pcm-iio will print fixed-width human readable values. While by default there will be no space after comma and raw integers.
What do you think?

[the remark this answers is not in the document]
Comment (ilia-kurakin) [reply pcm-iio.cpp:377]: I see, I think what you suggest is called "tsv" (tab-separated values). So I suggest adding `--csv-delimiter` optional argument that can be changed to support this human-readable output by choosing tab or any other character combination as a delimiter (e.g. `", "`). "human-readable" parameter still makes sense as an indicator to pick values format.

[the remark this answers is not in the document]
Comment (bigstream-dmytro) [reply pcm-iio.cpp]: I agree. Both arguments were added

File: pcm-iio.cpp [hunk 2]
```diff
@@ -333,11 +333,71 @@ vector<string> build_display(vector<struct iio_stacks_on_socket>& iios, vector<s
...
+    vector<string> current_row;
+    auto header = combine_stack_name_and_counter_names("Name");
+    header.insert(header.begin(), "Socket");
+    result.push_back(build_csv_row(header, csv_delimeter));
+    std::map<uint32_t,map<uint32_t,struct counter*>> v_sort;
+    //re-organize data collection to be row wise
+    size_t max_name_width = 0;
+    for (std::vector<struct counter>::iterator counter = ctrs.begin(); counter != ctrs.end(); ++counter) {
+        v_sort[counter->v_id][counter->h_id] = &(*counter);
+        max_name_width = std::max(max_name_width, counter->v_event_name.size());
```
Comment (opcm) [inline pcm-iio.cpp:357]: Windows MSVC compiler has a known issue with std::max . Compilation on appveyor fails:

c:\projects\pcm\pcm-iio_win\../pcm-iio.cpp(357): error C2589: '(': illegal token on right side of '::' [C:\projects\pcm\PCM-IIO_Win\pcm-iio-win.vcxproj]

tr y to fix it by putting std::max in brackets (std::max)(max_name_width, counter->v_event_name.size());

File: pcm-iio.cpp [hunk 3]
```diff
@@ -1021,18 +1081,121 @@ void print_PCIeMapping(const std::vector<struct iio_stacks_on_socket>& iios, con
...
+
+    while (argc > 1) {
+        argv++;
+        argc--;
+        std::string arg_value;
+        if (check_argument_equals(*argv, {"--help", "-h", "/h"})) {
+            print_usage(program);
+            exit(EXIT_FAILURE);
+        }
+        else if (extract_argument_value(*argv, {"-csv-delimeter", "/csv-delimeter"}, arg_value)) {
```
Comment (ilia-kurakin) [inline pcm-iio.cpp:1164]: Should be delim**i**ter rather than delim**e**ter.

**Review by ilia-kurakin — COMMENTED** [review]
I see 8 warnings that 8 lines in the patch add trailing whitespaces, please remove these whitespaces.

**Review by ilia-kurakin — APPROVED** [review]
@opcm, I think we're good to go

**Comment by opcm** [comment]
@bigstream-dmytro  thanks for the revised patches! @ilia-kurakin @ogbrugge thanks for the reviews.

**Pull request merged**